core/block/object/objectcreator/systemobjectreviser.go

```diff
@@ -0,0 +1,160 @@
+	"github.com/anyproto/anytype-heart/space"
+	"github.com/anyproto/anytype-heart/util/pbtypes"
+)
+
+var revisionKey = bundle.RelationKeyRevision.String()
+
+func (s *service) reviseSystemObjects(space space.Space, objects map[string]*types.Struct) {
+	marketRels, err := s.objectStore.ListAllRelations(addr.AnytypeMarketplaceWorkspace)
+	if err != nil {
+		log.Errorf("failed to get relations from marketplace space: %v", err)
+		return
+	}
+
+	marketTypes, err := s.listAllObjectTypes(addr.AnytypeMarketplaceWorkspace)
+	if err != nil {
+		log.Errorf("failed to get object types from marketplace space: %v", err)
+		return
+	}
+
+	for source, details := range objects {
+		if strings.HasPrefix(source, addr.BundledRelationURLPrefix) {
+			reviseSystemRelation(space, relationutils.RelationFromStruct(details), marketRels)
+		} else if strings.HasPrefix(source, addr.BundledObjectTypeURLPrefix) {
+			reviseSystemObjectType(space, details, marketTypes)
+		}
+	}
+}
+
+func (s *service) listAllObjectTypes(spaceId string) (map[string]*types.Struct, error) {
+	records, _, err := s.objectStore.Query(database.Query{
+		Filters: []*model.BlockContentDataviewFilter{
+			{
+				RelationKey: bundle.RelationKeyLayout.String(),
+				Condition:   model.BlockContentDataviewFilter_Equal,
+				Value:       pbtypes.Float64(float64(model.ObjectType_objectType)),
+			},
+			{
+				RelationKey: bundle.RelationKeySpaceId.String(),
+				Condition:   model.BlockContentDataviewFilter_Equal,
+				Value:       pbtypes.String(spaceId),
+			},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	details := make(map[string]*types.Struct, len(records))
+	for _, rec := range records {
+		id := pbtypes.GetString(rec.Details, bundle.RelationKeyId.String())
+		details[id] = rec.Details
+	}
+	return details, nil
+}
+
+func reviseSystemRelation(space space.Space, rel *relationutils.Relation, marketRels relationutils.Relations) {
+	marketRel := marketRels.GetModelByKey(rel.Key)
+	if marketRel == nil || !lo.Contains(bundle.SystemRelations, domain.RelationKey(rel.Key)) || marketRel.Revision <= rel.Revision {
+		return
+	}
+	details := buildRelationDiffDetails(marketRel, rel.Relation)
+	if len(details) != 0 {
+		if err := space.Do(rel.Id, func(sb smartblock.SmartBlock) error {
+			if ds, ok := sb.(basic.DetailsSettable); ok {
+				return ds.SetDetails(nil, details, false)
+			}
+			return nil
+		}); err != nil {
+			log.Errorf("failed to update system relation %s in space %s: %v", rel.Key, space.Id(), err)
+		}
+	}
+}
+
+func reviseSystemObjectType(space space.Space, objectType *types.Struct, marketTypes map[string]*types.Struct) {
+	marketType, found := marketTypes[pbtypes.GetString(objectType, bundle.RelationKeySourceObject.String())]
+	rawKey := pbtypes.GetString(objectType, bundle.RelationKeyUniqueKey.String())
+	uk, err := domain.UnmarshalUniqueKey(rawKey)
+	if !found || err != nil || !lo.Contains(bundle.SystemTypes, domain.TypeKey(uk.InternalKey())) ||
```

Comment:
maybe split if condition and move each check closely to places where corresponding variables appear?

```
marketType, found := marketTypes[pbtypes.GetString(objectType, bundle.RelationKeySourceObject.String())]
if !found {
  return
}
...
uk, err := domain.UnmarshalUniqueKey(rawKey)
if err != nil {
  return
}
...etc
```

Reply:
I have moved out checks on belonging to system objects to separate func, so it is used both for types and relations